src/drivers/telemetry/iridiumsbd/IridiumSBD.cpp

```diff
@@ -390,374 +455,347 @@ void IridiumSBD::sbdsession_loop(void)
+	if (!is_modem_ready()) {
+		VERBOSE_INFO("UPDATE SIGNAL QUALITY: MODEM NOT READY!");
+		return;
+	}
 
-	tx_buf_write_idx += buflen;
-	last_write_time = hrt_absolute_time();
+	write_at("AT+CSQ");
+	_new_state = SATCOM_STATE_CSQ;
+}
 
-	pthread_mutex_unlock(&tx_buf_mutex);
+void IridiumSBD::start_sbd_session(void)
+{
+	if ((_state != SATCOM_STATE_STANDBY) || (_new_state != SATCOM_STATE_STANDBY)) {
+		VERBOSE_INFO("CANNOT ENTER SBD SESSION STATE");
+		return;
 
-	return buflen;
+	} else {
+		VERBOSE_INFO("STARTING SBD SESSION");
+	}
+
+	if (!is_modem_ready()) {
+		VERBOSE_INFO("SBD SESSION: MODEM NOT READY!");
+		return;
+	}
+
+	if (_ring_pending) {
+		write_at("AT+SBDIXA");
+
+	} else {
+		write_at("AT+SBDIX");
+	}
+
+	_new_state = SATCOM_STATE_SBDSESSION;
+
+	pthread_mutex_lock(&_tx_buf_mutex);
+
+	_session_start_time = hrt_absolute_time();
 }
 
-ssize_t IridiumSBD::read(struct file *filp, char *buffer, size_t buflen)
+void IridiumSBD::start_test(void)
 {
-	PX4_DEBUG("READ: LEN %d, RX: %d RX END: %d", buflen, rx_msg_read_idx, rx_msg_end_idx);
+	if ((_state != SATCOM_STATE_STANDBY) || (_new_state != SATCOM_STATE_STANDBY)) {
+		PX4_INFO("CANNOT ENTER TEST STATE");
+		return;
+	}
 
-	if (rx_msg_read_idx < rx_msg_end_idx) {
-		size_t bytes_to_copy = rx_msg_end_idx - rx_msg_read_idx;
+	int res = read_at_command();
 
-		if (bytes_to_copy > buflen) {
-			bytes_to_copy = buflen;
+	if (res != SATCOM_RESULT_NA) {
+		PX4_WARN("SOMETHING WAS IN BUFFER");
+		printf("TEST RESULT: %d, LENGTH %d\nDATA:\n%s\nRAW DATA:\n", res, _rx_command_len, _rx_command_buf);
+
+		for (int i = 0; i < _rx_command_len; i++) {
+			printf("%d ", _rx_command_buf[i]);
 		}
 
-		memcpy(buffer, &rx_msg_buf[rx_msg_read_idx], bytes_to_copy);
+		printf("\n");
+	}
 
-		rx_msg_read_idx += bytes_to_copy;
+	if (!is_modem_ready()) {
+		PX4_WARN("MODEM NOT READY!");
+		return;
+	}
 
-		return bytes_to_copy;
+	if (strlen(_test_command) != 0) {
+		if ((strstr(_test_command, "AT") != nullptr) || (strstr(_test_command, "at") != nullptr)) {
+			PX4_INFO("TEST %s", _test_command);
+			write_at(_test_command);
+			_new_state = SATCOM_STATE_TEST;
+
+		} else {
+			PX4_WARN("The test command does not include AT or at: %s, ignoring it.", _test_command);
+			_new_state = SATCOM_STATE_STANDBY;
+		}
 
 	} else {
-		return -EAGAIN;
+		PX4_INFO("TEST DONE");
 	}
 }
 
-int IridiumSBD::ioctl(struct file *filp, int cmd, unsigned long arg)
+ssize_t IridiumSBD::write(struct file *filp, const char *buffer, size_t buflen)
 {
-	switch (cmd) {
-	case FIONREAD: {
-			int count = rx_msg_end_idx - rx_msg_read_idx;
-			*(int *)arg = count;
+	// general check if the incoming message would be too large (the buffer should not reset in that case)
+	if ((ssize_t)buflen > SATCOM_TX_BUF_LEN) {
+		return PX4_ERROR;
+	}
 
-			return OK;
+	pthread_mutex_lock(&_tx_buf_mutex);
+
+	// check if there is enough space for the incoming mavlink message
+	if (buflen == 6) {
+		if (*buffer == MAVLINK_PACKAGE_START) {
```

Comment:
I agree, that is why I added the check just after that one to catch anything which does not fit into that scheme.
However, in high latency mavlink 1 should be used because it assures that the HIGH_LATENCY2 message is 50 bytes. This is desired as for example rockblock bills in 50 byte increments.

Reply:
I adopted it to the way the protocol_splitter is doing it.

Mavlink 1 is not enforced but encouraged as the design of the HIGH_LATENCY2 stream was to create a 50 byte status message. Using Mavlink 2 this goal cannot be met due to the larger header. I added a note in the header file. Once the full high latency telemetry together with QGC is working there should be a documentation about it for example in the dev guide where this information should also be written.